File: files-app/src/main/java/l/files/fs/local/BasicFileTypeDetector.java
package l.files.fs.local;

import android.webkit.MimeTypeMap;

import com.google.common.net.MediaType;

import java.io.IOException;

import static com.google.common.io.Files.getFileExtension;
import static com.google.common.net.MediaType.OCTET_STREAM;

final class BasicFileTypeDetector extends LocalFileTypeDetector {

    public static final BasicFileTypeDetector INSTANCE = new BasicFileTypeDetector();

    private BasicFileTypeDetector() {
    }

    @Override
    protected MediaType detectRegularFile(
            LocalResource resource,
            LocalResourceStatus status) throws IOException {

        MimeTypeMap typeMap = MimeTypeMap.getSingleton();
        String ext = getFileExtension(resource.getName());
        String type = typeMap.getMimeTypeFromExtension(ext);
        if (type == null) {
            return OCTET_STREAM;
        }
        try {
            return MediaType.parse(type);
        } catch (IllegalArgumentException e) {
            return OCTET_STREAM;
        }
    }
}

File: files-app/src/main/kotlin/fs/local/BasicFileTypeDetector.kt
package l.files.fs.local

import android.webkit.MimeTypeMap

import com.google.common.net.MediaType

import l.files.fs.ResourceStatus

import com.google.common.net.MediaType.OCTET_STREAM
import org.apache.commons.io.FilenameUtils.getExtension
import l.files.fs.Path

private object BasicFileTypeDetector : LocalFileTypeDetector() {

    override fun detectRegularFile(status: ResourceStatus): MediaType {
        val typeMap = MimeTypeMap.getSingleton()
        val ext = getExtension(status.name)
        val type = typeMap.getMimeTypeFromExtension(ext)
        if (type == null) {
            return OCTET_STREAM
        }
        try {
            return MediaType.parse(type)
        } catch (e: IllegalArgumentException) {
            return OCTET_STREAM
        }
    }

}
